Feature/bulk data

boradol opened the pull request:
### PR Type
<!-- Please check the one that applies to this PR using "x". -->

- [ ] 버그수정[Bugfix]
- [x] 기능개발[Feature]
- [ ] 코드 스타일 변경[Code style update (formatting, local variables)]
- [ ] 리팩토링 [no functional changes, no api changes]
- [ ] 빌드 관련 변경
- [ ] 문서 내용 변경
- [ ] Other… Please describe:

---
### 요약(Summary)
1. 320개의 방탈출 테마 정보가 있는 Json 파일을 이용해 데이터를 넣는다.
2. ObjectMapper를 이용해서 Json형태를 객체형태로 변환한다.
3. 서로 연관관계를 생각하며 데이터를 넣는다.

---
### 상세 내용(Describe your changes)
- 데이터 넣을때 쿼리를 조회하면서 최대한 쿼리 발생이 적게 나오게 하려고 고민을 많이 했습니다.
- 도메인이나 dto역할에 맡게 책임을 부여하려고 하였습니다.
- 

---
### Issue Number or Link

---
### PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.  [Commit message convention](https://velog.io/@archivvonjang/Git-Commit-Message-Convention) 참고  (Ctrl + 클릭하세요.) 
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

boradol committed:
feat(ObjectMapperConfig): Bean ObjectMapper

boradol committed:
feat(Store): Save store by json file

1. ObjectMapper
2. Filter duplication data
3. SaveAll
4. Add Exception
5. json파일 일단 업로드 나중에 다시 삭제

boradol committed:
feat(Theme): Save Theme by json file

1. Save UnitPrice
2. Save Reservation
3. Separate DataUtils

boradol committed:
refactor(Store): Saperate DataUtils

1. Save UnitPrice
2. Save Reservation
3. Separate DataUtils

boradol committed:
refactor: Remove unnecessary annotations

boradol committed:
refactor(Theme): Classify existing objects for saving theme

boradol commented on src/main/java/com/project/escape/domain/reservation/Reservation.java:
```diff
@@ -13,17 +13,19 @@
     @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
-    @Column(nullable = false)
+    @Column
```
예약 일자넣는 것은 Json파일로 넣기 애매한 부분이 있었습니다. 고민후 넣어야 할 것 같습니다.
1. 예약가능한 시간은 정해져 있지만 예약 날짜같은 경우에는 무한정으로 만들어 줘야한다.
2. 정책이 필요한 부분 
  - 예를들어 일요일 0 시마다 그 다음주 일주일치의 예약이 오픈된다 -> 그럼 그때 그 다음주 일주일치 예약을 만드는게 좋은가?
  - 아니면, 서버는 예약가능한 시간의 테이블만 갖고있고, 클라이언트가 예약 날짜를 보내줘야 하는가?

boradol replied:
이 부분 멘토링 하면서 질문드릴려고 했는데요. 제가 조금 구조적으로 고민 더 해보겠습니다.

f-lab-kyle replied:
다른 사이트들을 보면 예약 가능한 날짜가 한달 까지만 오픈된걸 볼수 있는데요~
다 이유가 있지 않을까 싶네요 ~ 
- 미리 데이터를 만들어 두는것 : 비용 발생, 어디까지 만들어 둘것인가에 대한 모호함
- 서비스 측면 : 너무 미래의 예약건들이 있을때 관리하기 어려울수도

boradol commented on src/main/java/com/project/escape/domain/store/StoreDataUtils.java:
```diff
@@ -0,0 +1,50 @@
+
+@Component
+@RequiredArgsConstructor
+public class StoreDataUtils {
```
DataUtils : JsonFile을 readValue하여 해당 Dto로 반환하게끔 하는 메서드가 들어가 있음.
    |
    | - StoreDataUtils : 데이터 읽은것을 DTO로 변환 메서드, 저장메서드 포함
    |- ThemeDataUtils : 데이터 읽은것을 DTO로 변환 메서드, 저장메서드 포함

1. 패키지 구조
2. 도메인마다 DataUtils를 만드는 것에대한 고민
3. 만약 이 틀로.. 데이터를 계속 집어넣어야 한다면 문제점.

에 대해서

f-lab-kyle replied:
제가 잘 이해한건지 모르겠지만, StoreDataUtils, ThemeDataUtils 모두 테스트(dummy) 데이터를 만들기 위해 있는것 같은데요~ 
그렇다면 요건 쿼리로 처리해도 좋지 않을까 싶은 생각이 들었고요.

store나 theme를 cud하는건 아마도 컨트롤러에서 이루어지지 않을까 하는 생각도 들었어요. 가령, 어드민 같은 어떠한 요청에 의해서요.
그렇다면 서비스 같은 레이어로 두는게 적절하지 않을까 싶네요.

boradol replied:
제가 코멘트에 말을 하다 말았네요 ㅠㅠ 죄송합니다.
저도 멘토님이 이야기 해주신 부분이 이해가 잘 된건지 궁금합니다.

예를들어 어드민에서 더미데이터같이 넣을 수 있기 때문에 
도메인마다의 서비스 레이어로 두는것이 더 적절하다고 말씀해주신거죠?

boradol replied:
결론
- 저장하는 로직 -> xxxxxService
- Json to Dto로직 -> xxxxxJsonConverter

---

해결한 커밋
https://github.com/f-lab-edu/escape/pull/3/commits/d10ea99eba1b12b7c191a0a35a5c640b9787e385

boradol commented on src/main/java/com/project/escape/domain/store/StoreDataUtils.java:
```diff
@@ -0,0 +1,50 @@
+
+        return storeData.getHits().stream()
+                .distinct()
+                .filter(store -> !storeNames.contains(store.getStoreName()))
```
Json파일에 있는 Store정보와 
Table에 있는 Store정보를 비교하는 로직.
근데.... Json파일에서 유일하게 구분할 수 있는게 어쩔수 없이 name밖에 없어서 처리하게 되었습니다.

f-lab-kyle replied:
동일한 이름은 없겠죠? 유니크한 기준에 대해 고민도 필요해 보입니다.

boradol replied:
네네 동일한 이름이 없어 일단 넣게 되었습니다. 
Theme같은 경우에는 Json파일안에 refId가 있어서 이것을 primarykey로 넣게 되었어요.
근데.... Store같은 경우에는 없더라구요 ㅠㅠ

boradol commented on src/main/java/com/project/escape/domain/store/StoreDto.java:
```diff
@@ -0,0 +1,34 @@
+    public static StoreDto of(StoreDataDto.Hit request) {
+        return StoreDto.builder()
+                .name(request.getStoreName())
+                .address(new Address(mapLocation(request.getLocation()), request.getAddress()))
```
Embbed 를 사용하다보니 Dto에 Entity속성이 들어간다고 생각이 드는데 이런것도 Dto를 만들어서 넘겨줘야 하는지 코드를 짜면서 고민이 되었어요 🤔

f-lab-kyle replied:
객체의 생성 책임이 어디일지 고민이 필요해 보여요.
StoreDto를 만들기 위해 넘겨주는 파라미터에 따라 결정이 될수 있고요.
지금은 address라는 객체를 외부에서 주입시켜줘도 좋지 않을까 . 즉, 데이터 조합만 하고 생성 관련된 로직은 외부에 있는게 덜 의존하는 코드가 되지 않을까 생각되네요.

boradol replied:
해결한 커밋
https://github.com/f-lab-edu/escape/pull/3/commits/016ce953f265d87355b57e6587f4a3ff318c59fd
연관되어 해결한 커밋
https://github.com/f-lab-edu/escape/pull/3/commits/4748e6143877182f5cafb89244d1311fc49abea6

boradol commented on src/main/java/com/project/escape/domain/theme/ThemeDataUtils.java:
```diff
@@ -0,0 +1,72 @@
+
+    private Theme saveThemeData(Map<String, Store> storeMap, ThemeDto dto) {
+        Theme theme = ThemeDto.toEntity(dto, storeMap.get(dto.getStoreName()));
+        if (!themeRepository.existsById(theme.getId())) {
```
이렇게 분기처리 하지 않으면 쿼리상으
테마 Entity가 insert -> select -> update 되고 
modifiedDate도 변경이 일어나게 되더라고요.

Gpt에게 물어보니 
```
이런 경우에는 대부분 JPA가 엔티티를 관리하는 과정에서 발생하는 것입니다. JPA는 엔티티를 영속성 컨텍스트에 로드할 때 해당 엔티티의 변경 여부를 추적합니다. 그리고 트랜잭션을 커밋할 때 영속성 컨텍스트에 있는 엔티티의 변경사항을 데이터베이스에 반영합니다.

따라서, JPA는 엔티티를 로드한 후에는 해당 엔티티의 상태 변화를 추적하고, 트랜잭션이 커밋될 때 필요한 업데이트 쿼리를 실행합니다. 때문에 변경사항이 없더라도 JPA가 엔티티의 상태를 확인하고 업데이트 쿼리를 실행하는 것이 일반적입니다.

이런 동작을 피하기 위해서는 엔티티의 변경사항이 있는지를 명시적으로 확인하고, 변경이 없는 경우에는 업데이트 쿼리를 실행하지 않도록 제어해야 합니다. 예를 들어, 변경사항이 있는 경우에만 업데이트를 수행하도록 조건을 추가하거나, 변경사항이 없는 경우에는 업데이트를 수행하지 않도록 하여 이를 방지할 수 있습니다.

```

라고 하였고 저는 조건문을 넣게 되었습니다. 이것도 쿼리가 나가기는 하나 update문보다 낫지 않을까 생각이 되었습니다.

여기서 저는 영속성 컨테이너에서 변경이 일어나지 않으면  update하지 않는다고 생각했는데요.
왜 update가 일어나게 되는지 궁금해네요ㅠ

f-lab-kyle replied:
어느 부분이 바뀌는제 이것도 디버깅으로 확인 될것 같아요. 말씀하신데로 영속성 컨테이너값이 어떻게 바뀌는지도 실제로 확인해보시고요.

boradol commented on src/main/java/com/project/escape/domain/theme/ThemeDataUtils.java:
```diff
@@ -0,0 +1,72 @@
+    private Theme saveThemeData(Map<String, Store> storeMap, ThemeDto dto) {
+        Theme theme = ThemeDto.toEntity(dto, storeMap.get(dto.getStoreName()));
+        if (!themeRepository.existsById(theme.getId())) {
+            themeRepository.save(theme);
```
연관관계가 있어서 그런지 이쪽에서 원일 모를 쿼리들이 나타났습니다.

쿼리
```
Hibernate: select count(*) from theme t1_0 where t1_0.id=?

Hibernate: select null,t1_0.created_at,t1_0.description,t1_0.is_visible,t1_0.modified_at,t1_0.name,t1_0.playtime,t1_0.store_id from theme t1_0 where t1_0.id=? 
```
두번째 쿼리에 null이 들어가게 되었습니다.
null이 들어가는 이유는 store객체 때문이라고 Gpt에서 이야기를 하던데요.
사실 이부분이 잘 이해가 되지 않았습니다. ㅠㅠ

f-lab-kyle replied:
요건 코드 디버깅을 해보면 답이 나올것 같아요. jpa쪽.
아마 save를 하려고 할때 id 식별자 전략때문에 한번 체크하고 처리하려고 해서이지 않을까 생각되네요.

boradol commented on src/main/java/com/project/escape/global/exception/GlobalExceptionHandler.java:
```diff
@@ -8,6 +8,13 @@
 @RestControllerAdvice
 public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
+
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
```
지난번에 리뷰주신 내용을 반영해보았는데 제가 잘 이해한것일까요?

f-lab-kyle replied:
네!

boradol commented on src/test/java/com/project/escape/domain/store/StoreDataUtilsTest.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
```
이 아래 테스트부터는.. 데이터를 밀어넣는 동작만 수행하는 테스트코드같지 않은 것 밖에 못만들었네요.ㅜ

도메인 부분에 로직을 넣어 UnitTest도 넣어보고 싶었는데요. 완료하지 못해서 너무 아쉽습니다.ㅠㅠ
사실... 테스트코드를 작성해보긴 했는데 테스트 코드 짜는것 자체가 잘 안되는 것 같아요.
그냥 제가 아쉬워서 남깁니다.

f-lab-kyle replied:
테스트 코드는 사실 좀 다른 영역이라 생각이 들어요. (그냥 코드 작성하는 것과 다른 측면에서 고민이 되야 함)
왜 이 코드를 작성했는지, 어떤 인풋이 들어왔을때 어떤 아웃풋을 기대하는지, 해당 메소드의 역할과 책임이 무엇인지 등 유닛테스트를 작성할때 어떤걸 고려해봐야 할지에 대해서 공부가 필요합니다~

boradol reviewed (commented):
오늘도... 거의 멘토링 시작하기 전쯤에.. 부랴부랴 올리게 되었습니다.
뭔가 열심히는 하였는데... 결과가 나오지 않는 생산성 없는 개발자가 된거 같아 한심스럽습니다. 
아무튼 많은 내용이나 재밌는 주제가 없는 코드지만 그래도 많이 부족한 코드이니 리뷰 잘 부탁드립니다.🙏🏻

f-lab-kyle commented on src/main/java/com/project/escape/domain/reservation/ReservationService.java:
```diff
@@ -23,6 +23,11 @@ public class ReservationService {
 
+    @Transactional
+    public void saveAllReservations(List<Reservation> reservations) {
+        reservationRepository.saveAll(reservations);
```
ALL 류의 쿼리들은 늘 조심해야 할것 같아요. 성능에 직결되니..

boradol replied:
```suggestion
       reservations.forEach(reservationRepository::save);
```

이렇게 변경하면 아래처럼 소나린트가 추천을 해줍니다. 이 권고는 무시해도 되는 것이겠죠?
<img width="541" alt="image" src="https://github.com/f-lab-edu/escape/assets/58449849/f808fa45-36a9-4de3-8e70-9029c97c3e1d">

boradol replied:
`saveAll()`이 성능상 좋지 않을 수 있는지 찾아보았습니다.
---
일반적으로 JPA는 영속성 컨텍스트에 저장된 엔티티 객체의 상태 변화를 추적하여 변경된 부분만 UPDATE 쿼리를 생성합니다. 하지만 `saveAll()`을 사용하면 JPA는 저장하려는 모든 엔티티를 영속 상태로 변환하려고 시도합니다.

이때 만약 해당 엔티티가 이미 영속성 컨텍스트에 존재한다면, JPA는 불필요한 UPDATE 쿼리를 생성할 수 있습니다. 이는 성능 저하의 원인이 될 수 있습니다.

따라서 대량의 INSERT 작업에는 `saveAll()`이 효율적이지만, UPDATE 작업에는 `save()`를 개별적으로 사용하는 것이 더 나은 선택일 수 있습니다.

`saveAll()`과 같이 대량의 엔티티를 다룰 때, 불필요한 쿼리 생성으로 인해 성능 이슈가 발생할 수 있다는 의미입니다. 따라서 상황에 따라 적절한 메서드를 선택하는 것이 중요합니다.

boradol replied:
JPA에서 `saveAll()`과 `save()`의 성능을 비교
---

#### `saveAll()`

- 여러 개의 엔티티를 한 번에 저장할 수 있습니다.
- 한 번의 트랜잭션으로 처리되므로 성능이 좋습니다.
- 영속성 컨텍스트 관리가 효율적입니다.
- 단점은 한 엔티티에서 예외가 발생하면 전체 작업이 실패합니다.

#### `save()`

- 하나의 엔티티를 저장합니다.
- 매번 새로운 트랜잭션이 열리므로 오버헤드가 있습니다.
- 개별 엔티티에 대한 예외 처리가 용이합니다.
- 일반적으로 대량의 데이터를 한번에 저장해야 한다면 `saveAll()`을 사용하는 것이 효율적입니다. 
하지만 개별 - 엔티티에 대한 예외 처리가 중요하다면`save()`를 사용하는 것이 좋습니다.

성능 측면에서는 `saveAll()`이 `save()`보다 월등히 좋습니다. 한 번의 트랜잭션으로 처리되고 영속성 컨텍스트 관리도 효율적이기 때문입니다. 하지만 데이터의 크기와 예외 처리 요구사항에 따라 적절히 판단해야 합니다.

boradol replied:
select -> 몇 백만건이 들어올 것을 생각하여 페이지네이션 하기
insert -> 배치를 사용하여 chunk단위로 나눠서 하기.

f-lab-kyle commented on src/main/java/com/project/escape/domain/store/Location.java:
```diff
@@ -2,6 +2,22 @@
-    GYEONGGI, GANGWON, CHUNGCHEONG, JEOLLA, GYEONGSANG, JEJU
+    GYEONGGI, GANGWON, CHUNGCHEONG, JEOLLA, GYEONGSANG, JEJU, DEFAULT;
 
+    public static Location mapLocation(String location) {
```
이렇게 해도 상관은 없는데, Location.SEOUL 안에 데이터를 가지고 있어도 좋지 않을까 싶네요. 
가령, 서울의 도시를 출력하려면 지금대로라면 방법이 없을것 같아서요~

boradol replied:
데이터를 추천해주신대로 넣어보니 데이터 관리도 훨씬 더 나은것 같아요.
스위치문을 따로 만들지 않아도 되고 DEFAULT에 대한것도 추가해서 넣을 수 있어 좋아요.

그리고 이 Location도 Area로 변경하고 그 안의 데이터들의 값은 location으로 수정하였습니다.
의미상 Area가 Location보다 더 큰 범주라고 하네요.

boradol replied:
해결한 커밋
https://github.com/f-lab-edu/escape/pull/3/commits/bf6f940864e4640e8e1fdde80f0b55e46d0b6a05
https://github.com/f-lab-edu/escape/pull/3/commits/7bbf2b307074b00cbb802ea6a64d0b1e8aae7e9f

f-lab-kyle commented on src/main/java/com/project/escape/domain/store/StoreDataDto.java:
```diff
@@ -0,0 +1,34 @@
+@EqualsAndHashCode
+@NoArgsConstructor
+@AllArgsConstructor
+public class StoreDataDto {
```
DTO라면 그 목적에 맞는 네이밍이 더 좋을것 같고요.
data 라는 추상적인 이름은 사용 안하는게 좋을것 같아요

boradol replied:
해결한 커밋
https://github.com/f-lab-edu/escape/pull/3/commits/d10ea99eba1b12b7c191a0a35a5c640b9787e385

f-lab-kyle commented on src/main/java/com/project/escape/domain/store/StoreDataDto.java:
```diff
@@ -0,0 +1,34 @@
+        private String area;
+        private String address;
+        private String storeTel;
+        private Boolean storeIsopen;
```
`Boolean` vs  `boolean`

null을 표현할 상황이 있으려나요

boradol replied:
맞습니다. 
근데 getter 어노테이션을 쓰게되면 boolean같은 경우에는 isXXX이런식으로 변환되다 보니 
그 이후로 boolean보다는 Boolean으로 사용하게 된것 같아요.

이 Dto에서는 이 컬럼명으로 써야하는데요. 어떻게 하는것이 좋을까요?

f-lab-kyle commented on src/main/java/com/project/escape/global/utils/DataUtils.java:
```diff
@@ -0,0 +1,36 @@
+@Component
+@RequiredArgsConstructor
+public class DataUtils<T> {
+    private static final String EMPTY_STRING = "";
```
StringUtils.EMPTY

boradol replied:
해결한 커밋
https://github.com/f-lab-edu/escape/pull/3/commits/d10ea99eba1b12b7c191a0a35a5c640b9787e385

f-lab-kyle commented on src/main/java/com/project/escape/global/utils/DataUtils.java:
```diff
@@ -0,0 +1,36 @@
+        try {
+            return objectMapper.readValue(file, classType);
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
```
예외를 던지려면 에러로깅을 한곳에서 하는게 맞을것 같아요.
그리고 로그 어팬더를 사용하게 되는것과 이렇게 sysout로 출력하는것의 차이가 있어서요~ (로그 관리 측면)

boradol replied:
에러 처리하고 지운다는게 깜박하였네요.ㅜㅜ

boradol replied:
https://github.com/f-lab-edu/escape/pull/3/commits/d10ea99eba1b12b7c191a0a35a5c640b9787e385

f-lab-kyle commented on src/main/java/com/project/escape/global/utils/DataUtils.java:
```diff
@@ -0,0 +1,36 @@
+
+@Component
+@RequiredArgsConstructor
+public class DataUtils<T> {
```
어쩌면 convertUtil? 이라는 네이밍이 더 어울릴 것 같네요

boradol replied:
https://github.com/f-lab-edu/escape/pull/3/commits/d10ea99eba1b12b7c191a0a35a5c640b9787e385

f-lab-kyle commented on src/test/java/com/project/escape/domain/store/StoreDataUtilsTest.java:
```diff
@@ -0,0 +1,31 @@
+@ActiveProfiles(value = "local")
+@SpringBootTest
+class StoreDataUtilsTest {
+    @Autowired
```
이런 테스트는 CI 에서 돌기엔 다소 무리가 있죠.
bean을 전부 띄우다 보면 시간이 오래 걸려서요~
CI 에서 매번 돌릴지, 아니면 주기적으로 돌릴지 에 대한 package 를 분리해도 좋을것 같아요.

f-lab-kyle commented on src/test/java/com/project/escape/domain/theme/ThemeDataUtilsTest.java:
```diff
@@ -0,0 +1,32 @@
+    private ThemeDataUtils themeDataUtils;
+
+    @Test
+    @DisplayName("Json data를 읽어서 방탈출 테마 데이터가 저장된다. 방탈출 테마와 매핑되는 UnitPrice 정보와 Reservation 정보가 있으면 같이 저장된다.")
```
보통 displayName 또는 메서드 이름엔 given-when-then의 내용으로 작성하는 편입니다.
가령, 이 메소드는 
@DisplayName("저장하려는 테마의 개수와 저장된 테마의 개수가 같다.")
이렇게 작성할수 있겠네요.

그런데, 테스트의 목적이 좀 모호해 보이긴 합니다 ㅎ;

f-lab-kyle commented on src/main/java/com/project/escape/domain/store/StoreDto.java:
```diff
@@ -0,0 +1,34 @@
+    public static StoreDto of(StoreDataDto.Hit request) {
+        return StoreDto.builder()
+                .name(request.getStoreName())
+                .address(new Address(mapLocation(request.getLocation()), request.getAddress()))
```
메서드 네이밍이 모호한것 같아요.

getAddress? toAddress? generatedAddress?

f-lab-kyle reviewed (commented):
submit이 안눌려 있었네요;;

일단 이번 PR도 양이 많아 보입니다. 양을 줄이는 연습(작업을 작게 나누어 진행하는) 이 필요할것 같아요!
수고하셨습니다~

boradol committed:
refactor(JsonConvert): Fix JsonConvert

1. Rename StoreJsonConvert
2. Rename ThemeJsonConvert
3. Separate insert query in converter to service
4. Add dependency for test code

boradol committed:
refactor(Address): Add generate Address

boradol committed:
refactor(Score): Add generate Score

boradol committed:
refactor(Address): Rename Location -> Area & Add field locations

1. Area is Bigger than Location
2. Matching JSON Field Names and Entity Field Names

boradol committed:
refactor(Area): Fix mapArea

1. Caching AREA_MAP
2. Resolve DEFAULT

[Apply Review Link](https://github.com/f-lab-edu/escape/pull/3/files#r1545529801)

boradol commented:
질문이 있지만, 머지하도록 하겠습니다.
좋은 리뷰였습니다.
질문 멘토링때 할 수 있으면 궁금한거 여쭤보겠습니다. 감사합니다.
